### PR TITLE
Delete Task Executions

### LINKED ISSUE #1217: Add support to bulk-delete Task/Job executions
As a continuation from https://github.com/spring-cloud/spring-cloud-dataflow/issues/2600, we would want to add the options to delete the executions from the task executions list page.

- [x] Bulk-delete using the checkboxes
- [x] Task execution page: add delete action
- [x] Task definition page: list of executions: Bulk-delete using the checkboxes
- [x] Supply a date range similar to what we have in Audit list page; see:

image

### PR DESCRIPTION
**Pull request by oodamien**
- Task Executions list: add bulk delete action and delete action
- Task Execution: add delete action
- Task Definition - Executions list: add bulk delete action and delete action
- Add Task Executions Destroy modal

Resolves #1217

### CODE REVIEW

File: ui/src/app/tasks/task-definition/executions/task-definition-executions.component.ts [hunk 1]
```diff
@@ -113,4 +213,62 @@ export class TaskDefinitionExecutionsComponent implements OnInit, OnDestroy {
...
+   * by opening a confirmation modal dialog.
+   */
+  destroySelectedExecutions(page, params) {
+    const taskExecutions = page.items
+      .filter((item) => this.itemsSelected.indexOf(item.executionId) > -1);
+    this.destroyExecutions(taskExecutions, params);
+  }
+
+  /**
+   * Starts the destroy the {@link TaskExecution}s in parameter
```
Comment (ghillert) [inline ui/src/app/tasks/task-definition/executions/task-definition-executions.component.ts]: Maybe better: `Destroys the {@link TaskExecution}s that are provided as parameter. Will open a a confirmation modal dialog.`

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+        externalExecutionId: null
+      })
+    ];
+
+    beforeEach(() => {
+      component.open({ taskExecutions: mock });
+      fixture.detectChanges();
+    });
+
+    it('Should display a dedicate message', () => {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts]: dedicated

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    }).compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(TaskExecutionsDestroyComponent);
+    component = fixture.componentInstance;
+    notificationService.clearAll();
+  });
+
+  describe('1 tasks destroy', () => {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts]: Wording: `Destroy 1 task execution`

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+    it('Should call the service on validate destroy', () => {
+      const spy = spyOn(tasksService, 'destroyExecutions').and.callThrough();
+      const bt: HTMLElement = fixture.debugElement.query(By.css('#btn-destroy')).nativeElement;
+      bt.click();
+      fixture.detectChanges();
+      expect(spy).toHaveBeenCalledWith(mock);
+    });
+
+    it('Should display a message after destroy one task', (() => {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts]: destroying

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+      fixture.detectChanges();
+    });
+
+    it('Should display a dedicate message', () => {
+      const message: HTMLElement = fixture.debugElement.query(By.css('.modal-body')).nativeElement;
+      expect(message.textContent).toContain(`This action will destroy and delete the task execution`);
+      expect(message.textContent).toContain(`1`);
+    });
+
+    it('Should call the service on validate destroy', () => {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts]: Wording: `Should call the service after validating the button click to destroy the task execution`

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+      expect(spy).toHaveBeenCalledWith(mock);
+    });
+
+    it('Should display a message after destroy one task', (() => {
+      const bt: HTMLElement = fixture.debugElement.query(By.css('#btn-destroy')).nativeElement;
+      bt.click();
+      expect(notificationService.testSuccess[0]).toContain('1 task execution(s) destroyed.');
+    }));
+
+    it('Should close the modal after a success destroy', (() => {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts]: 'Should close the modal dialog after a success destruction'

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+      const bt: HTMLElement = fixture.debugElement.query(By.css('#btn-destroy')).nativeElement;
+      const spy = spyOn(bsModalRef, 'hide');
+      bt.click();
+      fixture.detectChanges();
+      expect(spy).toHaveBeenCalled();
+    }));
+
+  });
+
+  describe('Multiple tasks destroy', () => {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts]: Destruction of multiple task executions

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+      const tr1: HTMLElement = tr[0].nativeElement;
+      const tr2: HTMLElement = tr[1].nativeElement;
+      expect(tr.length).toBe(2);
+      expect(tr1.textContent).toContain(mock[0].executionId.toString());
+      expect(tr1.textContent).toContain(mock[0].taskName);
+      expect(tr2.textContent).toContain(mock[1].executionId.toString());
+      expect(tr2.textContent).toContain(mock[1].taskName);
+    });
+
+    it('Should call the service on validate destroy', () => {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts]: 'Should call the service upon validating the destruction of task executions'

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+      const spy = spyOn(tasksService, 'destroyExecutions').and.callThrough();
+      const bt: HTMLElement = fixture.debugElement.query(By.css('#btn-destroy')).nativeElement;
+      bt.click();
+      fixture.detectChanges();
+      expect(spy).toHaveBeenCalledWith(mock);
+    });
+
+    it('Should display a message after destroy 2 tasks', (() => {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts]: destroying

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+      const bt: HTMLElement = fixture.debugElement.query(By.css('#btn-destroy')).nativeElement;
+      bt.click();
+      expect(notificationService.testSuccess[0]).toContain('2 task execution(s) destroyed.');
+    }));
+
+    it('Should close the modal after a success destroy', (() => {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.spec.ts]: 'Should close the modal after a successful destruction'

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+   */
+  private ngUnsubscribe$: Subject<any> = new Subject();
+
+  /**
+   * Task Executions
+   */
+  taskExecutions: TaskExecution[];
+
+  /**
+   * Emit after destroy success
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.ts]: Emit events after the successful destruction of a task execution

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+  /**
+   * Initialize
+   */
+  open(args: { taskExecutions: TaskExecution[] }): Observable<any> {
+    this.taskExecutions = args.taskExecutions;
+    return this.confirm;
+  }
+
+  /**
+   * Submit destroy task(s)
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.ts]: Submit the destruction of task(s).

File: ui/src/app/tasks/task-executions/task-executions.component.html [hunk 13]
```diff
@@ -1,17 +1,17 @@
...
            class="table table-hover table-actions" id="taskExecutionsTable">
       <thead>
       <tr>
+        <th style="width: 30px" [dataflowAppRoles]="['ROLE_DESTROY']">
+          <app-master-checkbox (change)="changeCheckboxes()" *ngIf="form?.checkboxes" [dataflowAppRoles]="['ROLE_CREATE']"
+                               [items]="form.checkboxes"></app-master-checkbox>
+        </th>
         <th style="width: 140px">
           <app-sort id="sort-id" (change)="applySort($event)" [value]="'TASK_EXECUTION_ID'" [sort]="params">
             Execution Id
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions/task-executions.component.html]: Should call it `Task Execution Id` as there are also Job Execution ids etc, we should be more specific.

File: ui/src/app/tasks/task-executions/task-executions.component.ts [hunk 14]
```diff
@@ -13,6 +13,9 @@ import { AppError } from '../../shared/model/error.model';
 import { AuthService } from '../../auth/auth.service';
 import { TaskDefinition } from '../model/task-definition';
 import { TasksTabulationComponent } from '../components/tasks-tabulation/tasks-tabulation.component';
+import { TaskDefinitionsDestroyComponent } from '../task-definitions-destroy/task-definitions-destroy.component';
+import { TaskExecutionsDestroyComponent } from '../task-executions-destroy/task-executions-destroy.component';
+import { BsModalRef, BsModalService } from 'ngx-bootstrap';
 
 /**
  * Component that display the Task Executions.
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions/task-executions.component.ts]: displays

File: ui/src/app/tasks/task-executions/task-executions.component.ts [hunk 15]
```diff
@@ -44,6 +47,13 @@ export class TaskExecutionsComponent implements OnInit, OnDestroy {
   @ViewChild('tasksTabulation', { static: false })
   tasksTabulation: TasksTabulationComponent;
 
+  /**
+   * Current forms value
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions/task-executions.component.ts]: form

File: ui/src/app/tasks/task-executions/task-executions.component.ts [hunk 16]
```diff
@@ -60,19 +70,32 @@ export class TaskExecutionsComponent implements OnInit, OnDestroy {
    */
   context: any;
 
+  /**
+   * Contain a key application of each selected task executions
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions/task-executions.component.ts]: Contains

File: ui/src/app/tasks/task-executions/task-executions.component.ts [hunk 17]
```diff
@@ -118,6 +142,33 @@ export class TaskExecutionsComponent implements OnInit, OnDestroy {
...
+        action: 'destroy',
+        title: 'Destroy execution',
+        isDefault: false,
+        hidden: !this.authService.securityInfo.canAccess(['ROLE_DEPLOY']) || item.parentTaskExecutionId
+      },
+    ];
+  }
+
+  /**
+   * Tasks Actions
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions/task-executions.component.ts]: Task Actions

File: ui/src/app/tasks/task-executions/task-executions.component.ts [hunk 18]
```diff
@@ -118,6 +142,33 @@ export class TaskExecutionsComponent implements OnInit, OnDestroy {
...
+   */
+  executionsActions() {
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions/task-executions.component.ts]: executionActions

File: ui/src/app/tasks/task-executions/task-executions.component.ts [hunk 19]
```diff
@@ -183,6 +250,29 @@ export class TaskExecutionsComponent implements OnInit, OnDestroy {
     this.context.order = this.params.order;
     this.context.page = this.params.page;
     this.context.size = this.params.size;
+    this.context.itemsSelected = this.itemsSelected;
+  }
+
+  /**
+   * Update the list of selected checkbox
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions/task-executions.component.ts]: checkboxes

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.html [hunk 20]
```diff
@@ -0,0 +1,39 @@
+<div *ngIf="taskExecutions">
+  <div class="modal-header">
+    <h4 class="modal-title pull-left">Confirm Destroy Task Executions</h4>
+    <button type="button" class="close pull-right" aria-label="Close" (click)="cancel()">
+      <span aria-hidden="true">&times;</span>
+    </button>
+  </div>
+  <div class="modal-body" *ngIf="taskExecutions.length == 1">
+    <p>This action will destroy and delete the task execution <strong>{{ taskExecutions[0].executionId }}</strong>.
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.html]: `This action will remove the data of the task execution with the id `

File: ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.html [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+  <div class="modal-body" *ngIf="taskExecutions.length > 1">
+    <p>
+      This action will destroy the <strong>{{ taskExecutions.length }} task executions</strong> listed below.
+      Are you sure?
+    </p>
+    <br/>
+    <table id="table-tasks" class="table table-striped">
+      <thead>
+      <tr>
+        <th>Execution ID</th>
```
Comment (ghillert) [inline ui/src/app/tasks/task-executions-destroy/task-executions-destroy.component.html]: Task Execution ID

**Comment by sabbyanandan** [comment]
@ghillert: All the related backend changes are in; @oodamien's latest UI polishing should be good to go as well.

**Comment by ghillert** [comment]
Rebased, fixed some wording + roles. Merged with commit 6d0f9bdc6e4c1b4b5ccfb759104ac89b6e610792

**Pull request closed without merging**